File: backend/src/tenant-pricing/tenant-pricing.service.ts
import { Injectable, NotFoundException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { In, Repository } from 'typeorm';
import { TenantPricing } from '../common/entities/tenant-pricing.entity';
import { PricingModel } from '../common/entities/pricing-model.entity';
import { TenantsService } from '../tenants/tenants.service';
import { UpdateTenantPricingDto } from './dto/update-tenant-pricing.dto';

@Injectable()
export class TenantPricingService {
  constructor(
    @InjectRepository(TenantPricing)
    private readonly tenantPricingRepository: Repository<TenantPricing>,
    @InjectRepository(PricingModel)
    private readonly pricingRepository: Repository<PricingModel>,
    private readonly tenantsService: TenantsService
  ) {}

  async getByTenantId(tenantId: string) {
    const tenant = await this.tenantsService.getById(tenantId);
    if (!tenant) {
      throw new NotFoundException('Tenant not found');
    }

    const assignments = await this.tenantPricingRepository.find({ where: { tenantId } });
    return {
      tenantId,
      pricingIds: assignments.map((item) => item.pricingId)
    };
  }

  async upsert(tenantId: string, dto: UpdateTenantPricingDto) {
    const tenant = await this.tenantsService.getById(tenantId);
    if (!tenant) {
      throw new NotFoundException('Tenant not found');
    }

    const uniqueIds = Array.from(new Set(dto.pricingIds || []));
    if (uniqueIds.length > 0) {
      const pricing = await this.pricingRepository.find({
        where: { id: In(uniqueIds) }
      });
      if (pricing.length !== uniqueIds.length) {
        throw new NotFoundException('One or more pricing entries not found');
      }
    }

    await this.tenantPricingRepository.delete({ tenantId });
    if (uniqueIds.length === 0) {
      return { tenantId, pricingIds: [] };
    }

    const rows = uniqueIds.map((pricingId) =>
      this.tenantPricingRepository.create({ tenantId, pricingId })
    );
    await this.tenantPricingRepository.save(rows);
    return { tenantId, pricingIds: uniqueIds };
  }
}

File: backend-java/src/main/java/com/neria/manager/tenantpricing/TenantPricingService.java
package com.neria.manager.tenantpricing;

import com.neria.manager.common.entities.TenantPricing;
import com.neria.manager.common.repos.PricingModelRepository;
import com.neria.manager.common.repos.TenantPricingRepository;
import com.neria.manager.tenants.TenantsService;
import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.UUID;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;
import org.springframework.http.HttpStatus;
import org.springframework.transaction.annotation.Transactional;

@Service
public class TenantPricingService {
  private final TenantPricingRepository tenantPricingRepository;
  private final PricingModelRepository pricingRepository;
  private final TenantsService tenantsService;

  public TenantPricingService(
      TenantPricingRepository tenantPricingRepository,
      PricingModelRepository pricingRepository,
      TenantsService tenantsService) {
    this.tenantPricingRepository = tenantPricingRepository;
    this.pricingRepository = pricingRepository;
    this.tenantsService = tenantsService;
  }

  public TenantPricingResponse getByTenantId(String tenantId) {
    if (tenantsService.getById(tenantId) == null) {
      throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Tenant not found");
    }
    List<TenantPricing> assignments = tenantPricingRepository.findByTenantId(tenantId);
    return new TenantPricingResponse(
        tenantId, assignments.stream().map(TenantPricing::getPricingId).toList());
  }

  @Transactional
  public TenantPricingResponse upsert(String tenantId, TenantPricingUpdateRequest dto) {
    if (tenantsService.getById(tenantId) == null) {
      throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Tenant not found");
    }
    Set<String> unique = new HashSet<>(dto.pricingIds == null ? List.of() : dto.pricingIds);
    if (!unique.isEmpty()) {
      List<String> found =
          pricingRepository.findAllById(unique).stream().map(item -> item.getId()).toList();
      if (found.size() != unique.size()) {
        List<String> missing =
            unique.stream().filter(id -> !found.contains(id)).toList();
        throw new ResponseStatusException(
            HttpStatus.BAD_REQUEST, "Pricing entries not found: " + missing);
      }
    }
    tenantPricingRepository.deleteByTenantId(tenantId);
    tenantPricingRepository.flush();
    if (unique.isEmpty()) {
      return new TenantPricingResponse(tenantId, List.of());
    }
    List<TenantPricing> rows =
        unique.stream()
            .map(
                pricingId -> {
                  TenantPricing row = new TenantPricing();
                  row.setId(UUID.randomUUID().toString());
                  row.setTenantId(tenantId);
                  row.setPricingId(pricingId);
                  row.setCreatedAt(LocalDateTime.now());
                  return row;
                })
            .toList();
    tenantPricingRepository.saveAll(rows);
    return new TenantPricingResponse(tenantId, unique.stream().toList());
  }

  public record TenantPricingUpdateRequest(List<String> pricingIds) {}

  public record TenantPricingResponse(String tenantId, List<String> pricingIds) {}
}
